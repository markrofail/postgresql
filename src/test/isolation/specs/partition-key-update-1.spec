# Concurrency error from ExecUpdate and ExecDelete.

# Throw an error to indicate that the targeted row has been already moved to
# another partition in the case of concurrency where a session trying to
# update/delete a row that's locked for a concurrent update by the another
# session cause tuple movement to the another partition due update of partition
# key.

setup
{
  CREATE TABLE foo (a int, b text) PARTITION BY LIST(a);
  CREATE TABLE foo1 PARTITION OF foo FOR VALUES IN (1);
  CREATE TABLE foo2 PARTITION OF foo FOR VALUES IN (2);
  INSERT INTO foo VALUES (1, 'ABC');
}

teardown
{
  DROP TABLE foo;
}

session "s1"
setup		{ BEGIN; }
step "s1u"	{ UPDATE foo SET a=2 WHERE a=1; }
step "s1c"	{ COMMIT; }

session "s2"
step "s2u"	{ UPDATE foo SET b='EFG' WHERE a=1; }
step "s2d"	{ DELETE FROM foo WHERE a=1; }

permutation "s1u" "s1c" "s2u"
permutation "s1u" "s2u" "s1c"
permutation "s2u" "s1u" "s1c"

permutation "s1u" "s1c" "s2d"
permutation "s1u" "s2d" "s1c"
permutation "s2d" "s1u" "s1c"
