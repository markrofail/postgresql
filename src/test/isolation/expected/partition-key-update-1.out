Parsed test spec with 2 sessions

starting permutation: s1u s1c s2u
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s1c: COMMIT;
step s2u: UPDATE foo SET b='EFG' WHERE a=1;

starting permutation: s1u s2u s1c
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s2u: UPDATE foo SET b='EFG' WHERE a=1; <waiting ...>
step s1c: COMMIT;
step s2u: <... completed>
error in steps s1c s2u: ERROR:  tuple to be updated was already moved to another partition due to concurrent update

starting permutation: s2u s1u s1c
step s2u: UPDATE foo SET b='EFG' WHERE a=1;
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s1c: COMMIT;

starting permutation: s1u s1c s2d
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s1c: COMMIT;
step s2d: DELETE FROM foo WHERE a=1;

starting permutation: s1u s2d s1c
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s2d: DELETE FROM foo WHERE a=1; <waiting ...>
step s1c: COMMIT;
step s2d: <... completed>
error in steps s1c s2d: ERROR:  tuple to be updated was already moved to another partition due to concurrent update

starting permutation: s2d s1u s1c
step s2d: DELETE FROM foo WHERE a=1;
step s1u: UPDATE foo SET a=2 WHERE a=1;
step s1c: COMMIT;
